plutus-core/plutus-core/src/PlutusCore/Builtin/KnownType.hs

```diff
@@ -214,22 +221,18 @@ readKnownConstant mayCause val = asConstant mayCause val >>= oneShot \case
                 throwingWithCause _UnliftingError err mayCause
 {-# INLINE readKnownConstant #-}
 
--- See Note [Performance of KnownTypeIn instances].
--- We use @default@ for providing instances for built-in types instead of @DerivingVia@, because
--- the latter breaks on @m term@ (and for brevity).
--- | Haskell types known to exist on the PLC side.
--- Both the methods take a @Maybe cause@ argument to report the cause of a potential failure.
--- @cause@ is different to @term@ to support evaluators that distinguish between terms and values
--- (@makeKnown@ normally constructs a value, but it's convenient to report the cause of a failure
--- as a term). Note that an evaluator might require the cause to be computed lazily for best
--- performance on the happy path and @Maybe@ ensures that even if we somehow force the argument,
--- the cause stored in it is not forced due to @Maybe@ being a lazy data type.
--- Note that 'KnownTypeAst' is not a superclass of 'KnownTypeIn'. This is due to the fact that
--- polymorphic built-in types are only liftable/unliftable when they're fully monomorphized, while
--- 'toTypeAst' works for polymorphic built-in types that have type variables in them, and so the
--- constraints are completely different in the two cases and we keep the two classes apart
--- (there doesn't seem to be any cons to that).
-class uni ~ UniOf val => KnownTypeIn uni val a where
+{- Note [Cause of failure]
+'readKnown' and 'makeKnown' each take a @Maybe cause@ argument to report the cause of a potential
+failure. @cause@ is different to @val@ to support evaluators that distinguish between terms and
+values (@makeKnown@ normally constructs a value, but it's convenient to report the cause of a failure
+as a term). Note that an evaluator might require the cause to be computed lazily for best
+performance on the happy path and @Maybe@ ensures that even if we somehow force the argument,
+the cause stored in it is not forced due to @Maybe@ being a lazy data type.
+-}
+
+-- See Note [Performance of ReadKnownIn and MakeKnownIn instances].
+class uni ~ UniOf val => MakeKnownIn uni val a where
```

Comment:
No instances of these typeclasses ever fix `val`. What would happen if we did this?
```
class MakeKnownIn uni a where
  makeKnown :: forall val . (uni ~ UniOf val) => ...
```
I think so long as `makeKnown` gets inlined this is going to be no worse and it lets us get rid of a typeclass parameter.

Reply:
Okay, you'd need to add a `uni ~ UniOf val` constraint in a few places (e.g. kin the definition of `MakeKnown`), but that seems okay.

Reply:
I guess you're not gaining that much, it just seems somewhat interesting to me that really these instances only care about the *universe*, and you *can* use the same instance for whatever `val` you like, so long as it has that universe.

Reply:
Nope, I did miss something: the default implementation ultimately relies on `HasConstantIn` which does need to fix the term type. So you can't do the default instances without this. Leaving my comments just in case they're interesting anyway :shrug:

Reply:
I've been annoyed by this problem for years by now. Which is how #4172 emerged. It fixed the problem, but made the code slower and we reverted it. I'd been looking at it again today before I read your comments! I'm going to try it again now that we have a good inlining setup.

Reply:
> No instances of these typeclasses ever fix `val`

The `Opaque` instance does.

Reply:
Interesting. We weren't even inlining `makeKnown`/`readKnown` at that point, so perhaps it's less surprising that the dictionary-passing overhead didn't get optimized away.